server/search/facets.go

```
@@ -0,0 +1,202 @@
+// new values cannot be added to this data structure once it has been sorted
+func (f *SortedFacets) Add(tsCounts *tsApi.FacetCounts) error {
+	if tsCounts == nil || tsCounts.FieldName == nil {
+		return nil
+	}
+	if f.sorted {
+		return errors.New("Already initialized and sorted. No more inserts.")
+	}
+
+	if _, ok := f.facetAttrs[*tsCounts.FieldName]; !ok {
+		f.facetAttrs[*tsCounts.FieldName] = newFacetAttrs()
+	}
+	attrs := f.facetAttrs[*tsCounts.FieldName]
+
+	for i := 0; tsCounts.Counts != nil && i < len(*tsCounts.Counts); i++ {
+		c := (*tsCounts.Counts)[i]
+		if c.Value != nil {
+			attrs.addCount(*c.Value, c.Count)
+		}
+	}
+
+	attrs.addStats(tsCounts)
+	return nil
+}
+
+// GetFacetCount removes from priority queue and returns the value with highest count for the field if present
+// else returns nil, False
+func (f *SortedFacets) GetFacetCount(field string) (*FacetCount, bool) {
+	if !f.sorted {
+		f.sort()
+	}
+	if f.hasMoreFacets(field) {
+		if fc, err := f.counts[field].Pop(); err == nil {
+			return fc, true
+		} else {
+			log.Err(err)
+			return nil, false
+		}
+	}
+
+	return nil, false
+}
+
+// GetStats returns the computed stats for the faceted field
+func (f *SortedFacets) GetStats(field string) *api.FacetStats {
+	if attrs, ok := f.facetAttrs[field]; ok {
+		return attrs.stats
+	} else {
+		return nil
+	}
+}
+
+func (f *SortedFacets) hasMoreFacets(field string) bool {
+	if q, ok := f.counts[field]; ok {
+		return q.Len() > 0
+	}
+	return false
+}
+
+func (f *SortedFacets) initPriorityQueue(field string) {
+	if _, ok := f.counts[field]; !ok {
+		f.counts[field] = container.NewPriorityQueue[FacetCount](facetCountComparator)
+	}
+}
+
+// sort will queue up the collected unique facet counts in priority queue
+func (f *SortedFacets) sort() {
+	if f.sorted {
+		return
+	}
+
+	for field, attrs := range f.facetAttrs {
+		f.initPriorityQueue(field)
+		for _, fc := range attrs.counts {
+			if err := f.counts[field].Push(fc); err != nil {
```

if push throws an error should we return to debug what happened? I am concerned if we end up swallowing all these errors.

---

Removed the error from `pq.push()`